torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,396 @@
+// GraphExecutor is completely unaware of tracing or module parameters to keep the
+// tracing concerns separated.
+struct GraphExecutorImpl {
+
+  GraphExecutorImpl(std::shared_ptr<Graph> graph, bool optimize, bool symbolically_differentiable)
+  : graph(std::move(graph))
+  , optimize(optimize)
+  , symbolically_differentiable(symbolically_differentiable) {}
+  GraphExecutorImpl(std::shared_ptr<Graph> graph, bool optimize)
+  : graph(std::move(graph))
+  , optimize(optimize)
+  , symbolically_differentiable(isDifferentiable(*this->graph)) {}
+
+  // entry point where execution begins
+  variable_tensor_list run(variable_tensor_list inputs) {
+    // this is the fallback pathway, when we cannot differentiate
+    if(!optimize || (!symbolically_differentiable && gradientIsPossible(inputs))) {
+      auto & fb = getOrCreateAutogradFallback();
+      InterpreterState state(fb);
+      tensor_list outputs;
+      state.runOneStage(std::move(inputs), outputs);
+      // note: we never unwrapped inputs, because we want autograd to record the trace
+      return outputs;
+    }
+
+    // either we can symbolically differentiate, or we do not need a gradient.
+    // go down the route where we treat the inputs as tensors
+    // and fully optimize
+    auto & implementation = getOrCompile(inputs);
+    return implementation.run(std::move(inputs));
+  }
+
+private:
+
+  static bool gradientIsPossible(const variable_tensor_list & inputs) {
+    if (!autograd::GradMode::is_enabled()) {
+      return false;
+    }
+    for (const auto & tensor : inputs) {
+      if(tensor.defined() && Variable(tensor).requires_grad())
+        return true;
+    }
+    return false;
+  }
+  static bool isDifferentiable(Graph & g) {
+    for(auto n : g.nodes()) {
+      if(!jit::isDifferentiable(n))
+        return false;
+    }
+    return true;
+  }
+
+  static std::shared_ptr<Graph> copy(Graph & g) {
+    auto new_g = std::make_shared<Graph>();
+    std::unordered_map<Value*, Value*> value_map;
+    for(auto input : g.inputs()) {
+      value_map[input] = new_g->addInput()->copyMetadata(input);
+    }
+    for(auto node : g.nodes()) {
+      auto new_node = new_g->appendNode(new_g->createClone(node, [&](Value* v) {
+        return value_map.at(v);
+      }));
+      for(size_t i = 0; i < node->outputs().size(); ++i) {
+        value_map[node->outputs()[i]] = new_node->outputs()[i];
+        new_node->outputs()[i]->copyMetadata(node->outputs()[i]);
+      }
+    }
+    for(auto output : g.outputs()) {
+      new_g->registerOutput(value_map.at(output));
+    }
+    return new_g;
+  }
+
+  void runOptimization(std::shared_ptr<Graph> & graph, bool graphMustSupportVariables) {
+    EliminateDeadCode(graph);
+    CheckInplace(graph);
+    EliminateCommonSubexpression(graph);
+    if (!graphMustSupportVariables) {
+      // These optimizations can introduce operators like FusionGroup that
+      // do not work on variables
+      PeepholeOptimize(graph);
+      BatchMM(graph);
```

Comment:
BatchMM is a valid transformation on graphs that use Variables

Reply:
Well, it could be, but currently BatchMM checks for types and sizes. These won't be available on the size-agnostic Variable graphs. So we need to update that pass before it is valid to do.